### PR TITLE
feat(select): add compareWith Input for object value comparison

### PR DESCRIPTION
**Pull request by zakton5**
#### Short description of what this resolves:
Selects that use objects as values can end up with different object identities than the currently selected object, resulting in blank selects. More details provided in Angular link below.

#### Changes proposed in this pull request:
Leverage the same principle used in Angular's [compareWith](https://angular.io/docs/ts/latest/api/forms/index/SelectControlValueAccessor-directive.html)

**Ionic Version**: 3.x

**Fixes**: #6625 

Sorry about the multiple commits

### CODE REVIEW

File: src/components/select/select.ts [hunk 1]
```diff
@@ -451,9 +490,15 @@ export class Select extends BaseInput<any> implements OnDestroy {
...
-        option.selected = this.getValues().some(selectValue => {
-          return isCheckedProperty(selectValue, option.value);
-        });
+        if (this._compareWith) {
+          option.selected = this.getValues().some(selectValue => {
+            return this._compareWith(selectValue, option.value);
+          });
+        } else {
+          option.selected = this.getValues().some(selectValue => {
+            return isCheckedProperty(selectValue, option.value);
```
Comment (manucorporat) [inline src/components/select/select.ts]: what is  isCheckedProperty is the default, compare function? this way we don't have to duplicate code. does it make sense?

Comment (zakton5) [inline src/components/select/select.ts]: Yes, so I will add some additional logic into the `isCheckedProperty` function

**Comment by manucorporat** [comment]
>Is this what you meant? @manucorporat

yeah! now looks beautiful

File: src/components/select/select.ts [hunk 2]
```diff
@@ -188,6 +213,20 @@ export class Select extends BaseInput<any> implements OnDestroy {
...
+   */
+  @Input()
+  set compareWith(fn: (o1: any, o2: any) => boolean) {
+    if (typeof fn !== 'function') {
+      throw new Error(`compareWith must be a function, but received ${JSON.stringify(fn)}`);
+    }
+    this._compareWith = fn;
+  }
+
+  private _compareWith: (o1: any, o2: any) => boolean = isCheckedProperty;
```
Comment (manucorporat) [inline src/components/select/select.ts]: cc @zakton5

**Comment by manucorporat** [comment]
@zakton5 looks perfect now! but I prefer to wait until next version (3.5). 3.4 is scheduled for next wednesday.

**Comment by zakton5** [comment]
Alright, thanks for your help!

**Comment by manucorporat** [comment]
Thanks a lot for the PR!!

**Comment by zakton5** [comment]
@manucorporat No problem! Thanks for your help

**Pull request merged**